kafka-client/src/main/java/dev/responsive/kafka/internal/db/rs3/client/grpc/GrpcRS3Client.java

```diff
@@ -0,0 +1,234 @@
+      final RS3Grpc.RS3BlockingStub stub,
+      final RS3Grpc.RS3Stub asyncStub
+  ) {
+    this.channel = Objects.requireNonNull(channel);
+    this.stub = Objects.requireNonNull(stub);
+    this.asyncStub = Objects.requireNonNull(asyncStub);
+  }
+
+  public void close() {
+    channel.shutdownNow();
+  }
+
+  public static GrpcRS3Client connect(
+      final String target
+  ) {
+    final ChannelCredentials channelCredentials = InsecureChannelCredentials.create();
+    final ManagedChannel channel = Grpc.newChannelBuilder(target, channelCredentials)
+        .build();
+    final RS3Grpc.RS3BlockingStub stub = RS3Grpc.newBlockingStub(channel);
+    final RS3Grpc.RS3Stub asyncStub = RS3Grpc.newStub(channel);
+    return new GrpcRS3Client(channel, stub, asyncStub);
+  }
+
+  @Override
+  public CurrentOffsets getCurrentOffsets(final UUID storeId, final LssId lssId, final int pssId) {
+    final Rs3.GetOffsetsResult result;
+    try {
+      result = stub.getOffsets(Rs3.GetOffsetsRequest.newBuilder()
+          .setStoreId(uuidProto(storeId))
+          .setLssId(lssIdProto(lssId))
+          .setPssId(pssId)
+          .build());
+    } catch (final StatusRuntimeException e) {
+      throw new RS3Exception(e);
+    }
+    checkField(result::hasWrittenOffset, "writtenOffset");
+    checkField(result::hasFlushedOffset, "flushedOffset");
+    return new CurrentOffsets(
+        result.getWrittenOffset() == WAL_OFFSET_NONE
+            ? Optional.empty() : Optional.of(result.getWrittenOffset()),
+        result.getFlushedOffset() == WAL_OFFSET_NONE
+            ? Optional.empty() : Optional.of(result.getFlushedOffset())
+    );
+  }
+
+  @Override
+  public StreamSenderMessageReceiver<WalEntry, Optional<Long>> writeWalSegmentAsync(
+      final UUID storeId,
+      final LssId lssId,
+      final int pssId,
+      final Optional<Long> expectedWrittenOffset,
+      final long endOffset
+  ) {
+    final GrpcMessageReceiver<Rs3.WriteWALSegmentResult> resultObserver
+        = new GrpcMessageReceiver<>();
+    final var streamObserver = asyncStub.writeWALSegmentStream(resultObserver);
+    final var streamSender = new GrpcStreamSender<WalEntry, Rs3.WriteWALSegmentRequest>(
+        entry -> {
+          final var entryBuilder = Rs3.WriteWALSegmentRequest.newBuilder()
+              .setStoreId(uuidProto(storeId))
+              .setLssId(lssIdProto(lssId))
+              .setPssId(pssId)
+              .setEndOffset(endOffset)
+              .setExpectedWrittenOffset(expectedWrittenOffset.orElse(WAL_OFFSET_NONE));
+          if (entry instanceof Put) {
+            final var put = (Put) entry;
+            final var putBuilder = Rs3.WriteWALSegmentRequest.Put.newBuilder()
+                .setKey(ByteString.copyFrom(put.key()));
+            if (put.value().isPresent()) {
+              putBuilder.setValue(ByteString.copyFrom(put.value().get()));
+            }
+            entryBuilder.setPut(putBuilder.build());
+          }
```

Comment:
might make sense to implement a function `WalEntry#addTo(WriteWALSegmentRequestBuilder builder)` and implement it on each implementation of `WalEntry` instead of doing the check here

Reply:
I'd rather not put in the `WalEntry` type, as it's not aware of the specific format used by the underlying protocol. I added a helper method to this class.